In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import pandas as pd

In [3]:
X = pd.read_csv("X_processed.csv")

In [4]:
y = pd.read_csv("y_processed.csv")

In [5]:
X.head()

age   balance  day_of_week  duration  campaign  pdays  previous  \
0  0.129870  0.082382     0.900000  0.039854  0.000000    0.0       0.0   
1  0.233766  0.077524     0.733333  0.022367  0.016129    0.0       0.0   
2  0.259740  0.084733     0.500000  0.076861  0.016129    0.0       0.0   
3  0.298701  0.073058     0.100000  0.072794  0.016129    0.0       0.0   
4  0.116883  0.075436     0.300000  0.017893  0.000000    0.0       0.0   

   job_blue-collar  job_entrepreneur  job_housemaid  ...  month_jan  \
0                0                 0              0  ...          1   
1                1                 0              0  ...          0   
2                0                 0              0  ...          0   
3                0                 0              0  ...          0   
4                0                 0              0  ...          0   

   month_jul  month_jun  month_mar  month_may  month_nov  month_oct  \
0          0          0          0          0          0          0   
1          0          0          0          1          0          0   
2          1          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   month_sep  poutcome_other  poutcome_success  
0          0               1                 0  
1          0               1                 0  
2          0               1                 0  
3          0               1                 0  
4          0               1                 0  

[5 rows x 38 columns]

In [6]:
y.head()

y
0  0
1  0
2  0
3  0
4  0

In [7]:
y.value_counts(normalize=True)

y
0    0.5
1    0.5
Name: proportion, dtype: float64

```
import warnings
warnings.filterwarnings('ignore')
import pandas as pd
X = pd.read_csv("X_processed.csv")
y = pd.read_csv("y_processed.csv")
from sklearn.model_selection import GridSearchCV
import time
def grid_search(clf, params):
    gridsearch = GridSearchCV(clf, params, cv=5, n_jobs=-1)
    start_time = time.time()
    gridsearch.fit(X,y)
    end_time = time.time()  
    execution_time = end_time - start_time 
    print(gridsearch.best_params_)
    print(gridsearch.best_score_)
    print(f'{execution_time:.2f}')
from sklearn.model_selection import StratifiedKFold, cross_val_score
def crossval_recall(clf):
    skf = StratifiedKFold(n_splits=5, shuffle=True)
    start_time = time.time()
    score = cross_val_score(clf, X, y, cv=skf, n_jobs=-1, scoring='recall')
    end_time = time.time()  
    execution_time = end_time - start_time
    print(score)
    print(sum(score)/5)
    print(f'{execution_time:.2f}')
from sklearn.model_selection import RandomizedSearchCV
def random_search(clf, params, n_iter=100):
    randomsearch = RandomizedSearchCV(clf, params, cv=5, n_jobs=-1, n_iter=n_iter)
    start_time = time.time()
    randomsearch.fit(X,y)
    end_time = time.time()  
    execution_time = end_time - start_time 
    print(randomsearch.best_params_)
    print(randomsearch.best_score_)
    print(f'{execution_time:.2f}')
def crossval_accuracy(clf):
    skf = StratifiedKFold(n_splits=5, shuffle=True)
    start_time = time.time()
    score = cross_val_score(clf, X, y, cv=skf, n_jobs=-1, scoring='accuracy')
    end_time = time.time()  
    execution_time = end_time - start_time
    print(score)
    print(sum(score)/5)
    print(f'{execution_time:.2f}')
```

# Naive Bayes

In [9]:
X.var()

age                    0.024316
balance                0.000805
day_of_week            0.078636
duration               0.005153
campaign               0.001827
pdays                  0.015594
previous               0.000165
job_blue-collar        0.153749
job_entrepreneur       0.026398
job_housemaid          0.025324
job_management         0.173711
job_retired            0.065800
job_self-employed      0.032349
job_services           0.074619
job_student            0.031996
job_technician         0.137194
job_unemployed         0.033318
marital_married        0.245915
marital_single         0.217724
education_secondary    0.248699
education_tertiary     0.220320
default_yes            0.013797
housing_yes            0.249262
loan_yes               0.114357
contact_telephone      0.065069
month_aug              0.116231
month_dec              0.010939
month_feb              0.064336
month_jan              0.028984
month_jul              0.119394
month_jun              0.097866
month_ma

Low variance $\rightarrow$ use var_smoothing

In [9]:
from sklearn.model_selection import GridSearchCV

In [12]:
from sklearn.naive_bayes import GaussianNB

In [13]:
clf = GaussianNB()

In [31]:
import time

In [15]:
params = {
    'var_smoothing': [1e-9, 1e-7, 1e-5, 1e-3, 1e-1]
}

In [11]:
def grid_search(clf, params):
    gridsearch = GridSearchCV(clf, params, cv=5, n_jobs=-1)
    start_time = time.time()
    gridsearch.fit(X,y)
    end_time = time.time()  
    execution_time = end_time - start_time 
    print(gridsearch.best_params_)
    print(gridsearch.best_score_)
    print(f'{execution_time:.2f}')


In [ ]:
grid_search(clf, params)

In [17]:
params = {
    'var_smoothing': [1e-4, 0.0005, 1e-5, 0.00005,1e-6]
}

In [18]:
grid_search(clf, params)

{'var_smoothing': 0.0001}
0.6587475811893622
0.41


In [13]:
from sklearn.model_selection import StratifiedKFold, cross_val_score

In [21]:
clf = GaussianNB(var_smoothing=0.0001)

In [17]:
def crossval_recall(clf):
    skf = StratifiedKFold(n_splits=5, shuffle=True)
    start_time = time.time()
    score = cross_val_score(clf, X, y, cv=skf, n_jobs=-1, scoring='recall')
    end_time = time.time()  
    execution_time = end_time - start_time
    print(score)
    print(sum(score)/5)
    print(f'{execution_time:.2f}')

In [23]:
crossval_recall(clf)

[0.51606805 0.52646503 0.52835539 0.48438978 0.53024575]
0.5171047995807945
0.21


# Logistic Regression

In [25]:
from sklearn.linear_model import LogisticRegression

In [26]:
clf = LogisticRegression(random_state=42)

In [27]:
params = {
    'penalty': ['l1', 'l2', 'elasticnet', None],
    'C': [0.01, 0.1, 1],
    'tol': [1e-4, 1e-3, 1e-2]
}


In [28]:
grid_search(clf, params)

{'C': 0.01, 'penalty': None, 'tol': 0.0001}
0.7491075091501427
3.99


In [29]:
params={
    'penalty': [None],
    'C': [0.01, 0.001, 0.0001],
    'tol': [0.0001, 0.00001, 0.000001]
}

In [30]:
grid_search(clf, params)

{'C': 0.01, 'penalty': None, 'tol': 0.0001}
0.7491075091501427
2.63


In [31]:
clf = LogisticRegression(random_state=42, C=0.01, penalty=None, tol=0.0001)

In [32]:
crossval_recall(clf)

[0.79206049 0.79584121 0.78166352 0.79091769 0.80056711]
0.792210003344344
0.45


# Decision Trees

In [34]:
from sklearn.tree import DecisionTreeClassifier

In [35]:
clf = DecisionTreeClassifier(random_state=42)

In [36]:
params = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'max_depth': [5, 10, 20],  
    'min_samples_split': [2, 5, 10],  
    'min_samples_leaf': [1, 5, 10],  
    'max_leaf_nodes': [None, 20, 50]  
}

In [37]:
grid_search(clf, params)

{'criterion': 'gini', 'max_depth': 20, 'max_leaf_nodes': 50, 'min_samples_leaf': 10, 'min_samples_split': 2}
0.7422100667211876
29.73


In [38]:
params={
    'max_depth': [20, 25, 30, 35],
    'min_samples_leaf': [10, 15, 20, 25],
    'max_leaf_nodes':[50, 65, 70, 75]
}

In [39]:
grid_search(clf, params)

{'max_depth': 20, 'max_leaf_nodes': 50, 'min_samples_leaf': 15}
0.7461794187704174
7.96


In [40]:
clf = DecisionTreeClassifier(random_state=42, max_depth=20, max_leaf_nodes=50, min_samples_leaf=15)

In [41]:
crossval_recall(clf)

[0.86389414 0.88468809 0.83648393 0.86944182 0.86483932]
0.8638694597006544
0.32


# Random Forest

In [22]:
from sklearn.model_selection import RandomizedSearchCV

In [24]:
from sklearn.ensemble import RandomForestClassifier

'balance' over 'balanced_subsample' $\rightarrow$ dataset is not too big (<1e6 instances)

In [27]:
clf = RandomForestClassifier(random_state=42) # default n_estimators=100

In [47]:
params = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [5, 10],  
    'min_samples_split': [2, 5],  
    'min_samples_leaf': [1, 5]
}

In [21]:
def random_search(clf, params, n_iter=100):
    randomsearch = RandomizedSearchCV(clf, params, cv=5, n_jobs=-1, n_iter=n_iter)
    start_time = time.time()
    randomsearch.fit(X,y)
    end_time = time.time()  
    execution_time = end_time - start_time 
    print(randomsearch.best_params_)
    print(randomsearch.best_score_)
    print(f'{execution_time:.2f}')


In [ ]:
random_search(clf, params)

In [49]:
params = {
    'max_depth': [10, 15, 20],  
    'min_samples_split': [2, 1, None],  
    'min_samples_leaf': [5, 10, 15]
}

In [50]:
random_search(clf, params)

{'min_samples_split': 2, 'min_samples_leaf': 15, 'max_depth': 20}
0.742688108612977
22.83


In [29]:
clf = RandomForestClassifier(random_state=42, 
                             min_samples_leaf=15,
                             max_depth=20,
                             n_estimators=500
                            )

In [52]:
crossval_recall(clf)

[0.84877127 0.87712665 0.86862004 0.87890255 0.88279773]
0.8712436488760679
14.54


In [31]:
def crossval_accuracy(clf):
    skf = StratifiedKFold(n_splits=5, shuffle=True)
    start_time = time.time()
    score = cross_val_score(clf, X, y, cv=skf, n_jobs=-1, scoring='accuracy')
    end_time = time.time()  
    execution_time = end_time - start_time
    print(score)
    print(sum(score)/5)
    print(f'{execution_time:.2f}')

In [33]:
crossval_accuracy(clf)

[0.83223062 0.83506616 0.82939509 0.83309693 0.82931442]
0.8318206437946614
7.17


# XGBoost

In [9]:
from xgboost import XGBClassifier

In [65]:
clf = XGBClassifier(random_state=42, n_estimators=25)

In [67]:
params = {
    'booster': ['gbtree', 'dart'],  
    'eta': [0.01, 0.1],  # Taxa de aprendizado (menor = mais estável, maior = mais rápido)  
    'gamma': [0, 5],  # Regularização de divisão de nós (0 = mais splits, >0 = mais conservador)  
    'max_depth': [3, 10],  # Profundidade máxima da árvore (menor = menos overfitting)  
    'max_leaves': [10, 100],  # Número máximo de folhas por árvore (afeta complexidade)  
}

In [57]:
random_search(clf, params)

{'max_leaves': 10, 'max_depth': 3, 'gamma': 0, 'eta': 0.1, 'booster': 'gbtree'}
0.7695243713326809
67.59


In [58]:
params = {
    'eta': [0.1, 0.3, 0.5],  # Taxa de aprendizado (menor = mais estável, maior = mais rápido)  
    'gamma': [0, 1, 3],  # Regularização de divisão de nós (0 = mais splits, >0 = mais conservador)  
    'max_depth': [2, 3, 5],  # Profundidade máxima da árvore (menor = menos overfitting)  
    'max_leaves': [10, 20, 50],  # Número máximo de folhas por árvore (afeta complexidade)  
}

In [59]:
random_search(clf, params)

{'max_leaves': 10, 'max_depth': 2, 'gamma': 0, 'eta': 0.1}
0.7706556373370514
17.58


In [11]:
clf = XGBClassifier(random_state=42, 
                    n_estimators=100_000,
                    max_leaves=10,
                    max_depth=2,
                    gamma=0,
                    eta=0.1
                   )

In [13]:
crossval_recall(clf)

[0.81758034 0.80434783 0.82136106 0.79659413 0.80812854]
0.8096023807437319
194.22


In [18]:
crossval_accuracy(clf)

[0.81332703 0.83034026 0.8241966  0.80520095 0.80567376]
0.8157477197263224
185.33


# KNN

In [63]:
from sklearn.neighbors import KNeighborsClassifier

In [64]:
clf = KNeighborsClassifier()

In [65]:
params={
    'n_neighbors': [3,5,7,9],
    'weights': ['uniform', 'distance'],
    'metric': ['chebyshev', 'manhattan', 'euclidian'],
    'algorithm': ['auto', 'ball_tree', 'kd_tree']
}

In [66]:
random_search(clf, params)

{'weights': 'uniform', 'n_neighbors': 9, 'metric': 'manhattan', 'algorithm': 'ball_tree'}
0.6298235217882887
86.18


In [67]:
params={
    'n_neighbors': [9, 11, 13, 15, 17, 19, 21, 23, 25],
    'metric': ['manhattan'],
    'algorithm': ['ball_tree']
}

In [68]:
random_search(clf, params)

{'n_neighbors': 15, 'metric': 'manhattan', 'algorithm': 'ball_tree'}
0.6405074027895088
22.32


In [69]:
clf = KNeighborsClassifier(
    n_neighbors=15,
    metric='manhattan',
    algorithm='ball_tree'
)

In [70]:
crossval_recall(clf)

[0.61720227 0.62854442 0.64933837 0.64049196 0.62759924]
0.6326352536783313
2.66


# SVM

In [2]:
from sklearn.svm import SVC

In [5]:
clf = SVC(random_state=42)

In [7]:
params = [
    {'C': [0.1, 1, 5], 'kernel': ['linear']},  # Linear não precisa de gamma nem degree
    {'C': [0.1, 1, 5], 'kernel': ['poly'], 'degree': [2, 3], 'gamma': ['scale', 'auto']},
    {'C': [0.1, 1, 5], 'kernel': ['rbf', 'sigmoid'], 'gamma': ['scale', 'auto']}
]

In [9]:
random_search(clf, params)

{'kernel': 'linear', 'C': 5}
0.7576137678924952
218.40


In [11]:
params={
    'kernel':['linear'],
    'C':[2, 3, 5, 6, 7]
}

In [13]:
random_search(clf, params)

{'kernel': 'linear', 'C': 5}
0.7576137678924952
40.15


In [15]:
clf = SVC(random_state=42,
          kernel='linear',
          C=5
         )

In [17]:
crossval_recall(clf)

[0.79111531 0.79678639 0.81379962 0.79470199 0.81096408]
0.8014734786364375
6.81


# MLP

In [19]:
from sklearn.neural_network import MLPClassifier

In [21]:
clf = MLPClassifier(random_state=42)

Hidden layer sizes:
$$
HLS=\sqrt{38_{features}\cdot2_{classes}}\approx9
$$

In [37]:
params={
    'hidden_layer_sizes': [8, 9, 10],
    'activation': ['logistic', 'relu'],
    'solver': ['lbfgs', 'sgd', 'adam'],
    'alpha': [0.1, 0.01, 0.001],
    'learning_rate': ['constant', 'invscaling', 'adaptive']
}

In [39]:
random_search(clf, params)

{'solver': 'adam', 'learning_rate': 'adaptive', 'hidden_layer_sizes': 8, 'alpha': 0.001, 'activation': 'logistic'}
0.751943316038558
368.14


In [41]:
params={
    'hidden_layer_sizes': [5, 6, 7, 8],
    'activation': ['logistic'],
    'alpha': [0.001, 0.0001, 0.00001],
    'learning_rate': ['adaptive']
}

In [43]:
random_search(clf, params)

{'learning_rate': 'adaptive', 'hidden_layer_sizes': 8, 'alpha': 1e-05, 'activation': 'logistic'}
0.7522268699137943
68.42


In [55]:
clf = MLPClassifier(random_state=42,
                    learning_rate='adaptive', 
                    hidden_layer_sizes=8, 
                    alpha=1e-05, 
                    activation='logistic'
                   )

In [57]:
crossval_recall(clf)

[0.78166352 0.79773157 0.79489603 0.81078524 0.8194707 ]
0.8009094111987238
10.45


|         Model         | Avg Accuracy | Avg Recall |
|:---------------------:|:------------:|:----------:|
|      Naive Bayes      |      66%     |     52%    |
|  Logistic Regression  |      75%     |     79%    |
|     Decision Tree     |      75%     |     86%    |
|     Random Forest     |      83%     |     87%    |
|        XGBoost        |      81%     |     81%    |
|          KNN          |      64%     |     63%    |
|          SVM          |      76%     |     80%    |
| Multilayer Perceptron |      75%     |     80%    |

# Top Three
1. Random Forest
- Parameters: {'min_samples_split': 2, 'min_samples_leaf': 15, 'max_depth': 20, 'n_estimators': 500}
2. XGBoost
- Parameters: {'max_leaves': 10, 'max_depth': 2, 'gamma': 0, 'eta': 0.1, 'n_estimators': 100000}
3. SVM
- Parameters: {'kernel': 'linear', 'C': 5}